pypi.eclass: optimizations

mgorny opened the pull request:
These changes make the eclass run twice as fast for the "common" use case.

Technically, there are still some possible optimizations:
- using `_pypi_translate_version` in wheel functions
- "caching" normalized project name (right now we do it twice, for `S` and as part of `pypi_sdist_url`)
- "caching" betwen `pypi_wheel_name` and `pypi_wheel_url`

That said, I'm not sure if there's a point in adding more complexity, given these functions aren't used all that often.

mgorny commented:
I forgot to remove the last subshell from the default code path. The rough result is almost 3 times faster than the original.

eli-schwartz commented on eclass/pypi.eclass:
```diff
@@ -63,6 +63,21 @@ _PYPI_ECLASS=1
+# via _PYPI_NORMALIZED_NAME variable.
+_pypi_normalize_name() {
+	local name=${1}
+	local shopt_save=$(shopt -p extglob)
+	shopt -s extglob
```
You can save another fork:
```bash
local shopt_save=
if ! shopt -p extglob; then
    shopt_save='shopt -u extglob'
    shopt -s extglob
fi
do_stuff
$shopt_save
```

mgorny replied:
Nice idea. I'll try later, when I have free CPU.

mgorny replied:
Thanks, this is great! 550% speedup ;-).

mgorny commented:
I've added the shopt optimization and replaced the results in commit messages with 1/n that make much more sense.

The final result is that the eclass is 16 times faster for the most common case.

thesamesam reviewed (approved):
LGTM, just make sure you diff metadata before/after for unexpected SRC_URI changes

fantastic work!

thesamesam commented:
btw, could you add Bug: tags to each commit?

ionenwks reviewed (commented):
Haven't checked why, but with the changes:
```
>>> Emerging (1 of 1) dev-python/PyQt6-WebEngine-6.5.0::gentoo
 * PyQt6_WebEngine-6.5.0.tar.gz BLAKE2B SHA512 size ;-) ...                                                  [ ok ]
>>> Unpacking source...
>>> Unpacking PyQt6_WebEngine-6.5.0.tar.gz to /tmp/portage/dev-python/PyQt6-WebEngine-6.5.0/work
>>> Source unpacked in /tmp/portage/dev-python/PyQt6-WebEngine-6.5.0/work
 * ERROR: dev-python/PyQt6-WebEngine-6.5.0::gentoo failed (prepare phase):
 *   The source directory '/tmp/portage/dev-python/PyQt6-WebEngine-6.5.0/work/PyQt6_WebEngine-' doesn't exist
```

ionenwks commented on eclass/pypi.eclass:
```diff
@@ -221,12 +258,27 @@ pypi_wheel_url() {
+
+	if [[ ${PYPI_NO_NORMALIZE} ]]; then
+		_pypi_sdist_url --no-normalize "${PYPI_PN}" "${_PYPI_TRANSLATED_VERSION}"
+		S="${WORKDIR}/${PYPI_PN}-${version}"
```
Well, wrt previous comment I see now:
```suggestion
		S="${WORKDIR}/${PYPI_PN}-${_PYPI_TRANSLATED_VERSION}"
```

mgorny replied:
Good catch, thanks!
